lib/tesla/multipart.ex

```diff
@@ -0,0 +1,160 @@
+defmodule Tesla.Multipart do
+  @moduledoc """
+  Multipart functionality.
+
+  Example:
+  ```ex
+      mp =
+        Multipart.new
+        |> Multipart.add_content_type_param("charset=utf-8")
+        |> Multipart.add_field("field1", "foo")
+        |> Multipart.add_field("field2", "bar", headers: [{:"Content-Id", 1}, {:"Content-Type", "text/plain"}])
+        |> Multipart.add_file("test/tesla/multipart_test_file.sh")
+        |> Multipart.add_file("test/tesla/multipart_test_file.sh", name: "foobar")
+
+        response = client.post(url, mp)
+   ```
+  """
+
+  @boundary_chars "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789" |> String.split("")
+
+  @type part_stream :: IO.Stream.t | File.Stream.t
+  @type part_value :: String.t | part_stream
+
+  defstruct [
+    parts: [],
+    boundary: nil,
+    content_type_params: []
+  ]
+
+  @type t :: %__MODULE__ {
+    parts: list(Part.t),
+    boundary: String.t,
+    content_type_params: [String.t]
+  }
+
+  defmodule Part do
+    defstruct [
+      body: nil,
+      dispositions: [],
+      headers: [],
+    ]
+
+    @type t :: %__MODULE__ {
+      body: String.t,
+      headers: Keyword.t,
+      dispositions: Keyword.t,
+    }
+  end
+
+  @doc """
+  Create a new Multipart struct to be used for a request body.
+  """
+  @spec new() :: t
+  def new do
+    %__MODULE__{boundary: unique_string(32)}
+  end
+
+  @doc """
+  Add a parameter to the multipart content-type.
+  """
+  @spec add_content_type_param(t, String.t) :: t
+  def add_content_type_param(%__MODULE__{} = mp, param) do
+    %{mp | content_type_params: mp.content_type_params ++ [param]}
+  end
+
+  @doc """
+  Add a field part.
+  """
+  @spec add_field(t, String.t, part_value, Keyword.t) :: t
+  def add_field(%__MODULE__{} = mp, name, value, opts \\ []) do
+    {headers, opts} = Keyword.pop_first(opts, :headers, [])
+
+    part = %Part{
+      body: value,
+      headers: headers,
+      dispositions: [{:name, name}] ++ opts
+    }
+    %{mp | parts: mp.parts ++ [part]}
+  end
+
+  @doc """
+  Add a file part. The file will be streamed.
```

Comment:
same as add_field, also I would add additional example with `:detect_content_type` to moduledoc